### PR TITLE
feat(Themes)!: Initialize Core + Platform Themes

### PR DESCRIPTION
**Pull request by codecaaron**
### Overview

Monolith PR: https://github.com/codecademy-engineering/Codecademy/pull/24393

This introduces separate themes for core and platform through gamut-styles and fixes a bunch of the existing behavior to allow for themes to be extended.

This also requires us to update the way we declare the global emotion theme, removing the implicit Core declaration.   This ensures consuming apps will be allowed to declare their own, and all system props types will be derived from the explicit declaration.

As these themes are fully self contained now I have added snapshot tests to ensure that we do not accidentally lose tokens that might break behavior silently.

#### Core Theme
- It includes our basic palette and general sets of tokens.
- Gamut components are only aware of this theme and gamut will compile with reference to this themes types alone.
- Gamut components / system props while within gamut are limited to this, to ensure interoperability between extensions of this base theme.

#### Platform Theme
- Adds a set of net new colors for the learning environment.
- Cannot be consumed within Gamut (shared color mode aliases will work however :)

#### `createTheme`
- Return object is just theme (all methods and extra stuff are on it under _ prefixed keys.  e. g. theme._getColorValue(x)
- New utility `flattenScale` The recursively flattens an object to kebab case keys. gray[200] => gray-200  and you may now pass a deeper object to some methods and receive a strongly typed map back.
```tsx
const object = {
  gray: {
    '200': '#eeeeee',
     '600': '#666666',
  }
}

// flattentScale(object) 
const output = {
  'gray-200': '#eeeeee',
  'gray-600': '#666666',
}
```

- Ensures that simpler types now support extension.  A theme that is returned by `createTheme().build()` can be extended again with createTheme with no loss.
```tsx
// valid
const base = createTheme(x).addScale().build();
// also valid but extra!
const extended = createTheme(base).addScale().build();
```

#### GamutProvider
-  Passing theme is now required by GamutProvider 
- The theme passed to gamut provider will automatically set global variables from the `_variables` key.

#### Modes
- `Background` will now cycle through any arbitrary amount of color modes without specifically requiring their names.
- Background now is agnostic to the theme that is passed using only the methods / tokens available on the supplied context to check contrast.
- Exports many new types for colors `Colors` + `ColorModes` etc.
- Adds new hooks
   - `useCurrentMode(modeOverride?: ColorMode): ColorMode` Optionally takes an override mode and returns the value of either the override or the current context.
   - `useColorMode()`:  Returns all relevant mode keys and _getColorValue. 
- Color Modes now do not have unnecessary depth on the them and are now top level keys.  This wasn't ergonomic for utilities or general use.
```tsx
//before
const theme = {
  colorModes: { 
      modes: { light: {}, dark: {} }, 
      active: 'light' 
  }
}

// after 
const theme = {
  modes: { light: {}, dark: {} },
  mode: 'light',
}
```

#### Token clean up
- Consolidates grayscale to a limited set following the rules in the figma doc.
- Moves all original colors to the deprecated colors folder.  Creates new swatches and aliases to decouple the sets.
- Adds new color swatches for shadows and other effects that are rarely used with more consistent key names. 
- Many tokens have been cleaned up as they are no longer used or serve no purpose.  See comments for explanations.

### PR Checklist

- [x] Related to designs: [Figma](https://www.figma.com/file/QplOE1d1rHbQLu0PtvsWLX/Colors-%E2%9C%85?node-id=487%3A2)
- [x] Related to JIRA ticket: https://codecademy.atlassian.net/browse/GM-174 https://codecademy.atlassian.net/browse/GM-228
- [x] I have run this code to verify it works
- [x] This PR includes unit tests for the code change

### CODE REVIEW

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-illustrations@0.13.5-alpha.eeb01a.0<br/>@codecademy/gamut-kit@0.3.77-alpha.eeb01a.0<br/>@codecademy/gamut-labs@15.5.1-alpha.eeb01a.0<br/>@codecademy/gamut-styles@9.8.1-alpha.eeb01a.0<br/>@codecademy/gamut-tests@2.3.28-alpha.eeb01a.0<br/>@codecademy/gamut@32.11.1-alpha.eeb01a.0<br/>@codecademy/konamimojisplosion@0.2.28-alpha.eeb01a.0<br/>@codecademy/markdown-overrides@0.5.40-alpha.eeb01a.0<br/>@codecademy/styleguide@36.13.3-alpha.eeb01a.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6092bcaf325b1d0ef7ecc27c--gamut-preview.netlify.app>https://6092bcaf325b1d0ef7ecc27c--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6092bcaf325b1d0ef7ecc27c>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6099af52d210450090fc0cfb--gamut-preview.netlify.app>https://6099af52d210450090fc0cfb--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6099af52d210450090fc0cfb>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@4.12.1-alpha.0a1301.0<br/>@codecademy/gamut-illustrations@0.14.2-alpha.0a1301.0<br/>@codecademy/gamut-kit@0.3.89-alpha.0a1301.0<br/>@codecademy/gamut-labs@15.8.2-alpha.0a1301.0<br/>@codecademy/gamut-patterns@0.1.2-alpha.0a1301.0<br/>@codecademy/gamut-styles@9.8.2-alpha.0a1301.0<br/>@codecademy/gamut-tests@2.3.29-alpha.0a1301.0<br/>@codecademy/gamut@32.15.3-alpha.0a1301.0<br/>@codecademy/konamimojisplosion@0.2.29-alpha.0a1301.0<br/>@codecademy/markdown-overrides@0.5.41-alpha.0a1301.0<br/>@codecademy/styleguide@37.2.3-alpha.0a1301.0<br/>@codecademy/variance@0.11.3-alpha.0a1301.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6099b0f8e1603100a0684ef1--gamut-preview.netlify.app>https://6099b0f8e1603100a0684ef1--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6099b0f8e1603100a0684ef1>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@4.12.1-alpha.042827.0<br/>@codecademy/gamut-illustrations@0.14.2-alpha.042827.0<br/>@codecademy/gamut-kit@0.3.89-alpha.042827.0<br/>@codecademy/gamut-labs@15.8.2-alpha.042827.0<br/>@codecademy/gamut-patterns@0.1.2-alpha.042827.0<br/>@codecademy/gamut-styles@9.8.2-alpha.042827.0<br/>@codecademy/gamut-tests@2.3.29-alpha.042827.0<br/>@codecademy/gamut@32.15.3-alpha.042827.0<br/>@codecademy/konamimojisplosion@0.2.29-alpha.042827.0<br/>@codecademy/markdown-overrides@0.5.41-alpha.042827.0<br/>@codecademy/styleguide@37.2.3-alpha.042827.0<br/>@codecademy/variance@0.11.3-alpha.042827.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6099b690d0c62600c0921b32--gamut-preview.netlify.app>https://6099b690d0c62600c0921b32--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6099b690d0c62600c0921b32>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@4.12.1-alpha.bcb3fb.0<br/>@codecademy/gamut-illustrations@0.14.2-alpha.bcb3fb.0<br/>@codecademy/gamut-kit@0.3.89-alpha.bcb3fb.0<br/>@codecademy/gamut-labs@15.8.2-alpha.bcb3fb.0<br/>@codecademy/gamut-patterns@0.1.2-alpha.bcb3fb.0<br/>@codecademy/gamut-styles@9.8.2-alpha.bcb3fb.0<br/>@codecademy/gamut-tests@2.3.29-alpha.bcb3fb.0<br/>@codecademy/gamut@32.15.3-alpha.bcb3fb.0<br/>@codecademy/konamimojisplosion@0.2.29-alpha.bcb3fb.0<br/>@codecademy/markdown-overrides@0.5.41-alpha.bcb3fb.0<br/>@codecademy/styleguide@37.2.3-alpha.bcb3fb.0<br/>@codecademy/variance@0.11.3-alpha.bcb3fb.0<br/></pre>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@4.12.1-alpha.5d5070.0<br/>@codecademy/gamut-illustrations@0.14.2-alpha.5d5070.0<br/>@codecademy/gamut-kit@0.3.89-alpha.5d5070.0<br/>@codecademy/gamut-labs@15.8.2-alpha.5d5070.0<br/>@codecademy/gamut-patterns@0.1.2-alpha.5d5070.0<br/>@codecademy/gamut-styles@9.8.2-alpha.5d5070.0<br/>@codecademy/gamut-tests@2.3.29-alpha.5d5070.0<br/>@codecademy/gamut@32.15.3-alpha.5d5070.0<br/>@codecademy/konamimojisplosion@0.2.29-alpha.5d5070.0<br/>@codecademy/markdown-overrides@0.5.41-alpha.5d5070.0<br/>@codecademy/styleguide@37.2.3-alpha.5d5070.0<br/>@codecademy/variance@0.11.3-alpha.5d5070.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://609a0546634480569d2af85f--gamut-preview.netlify.app>https://609a0546634480569d2af85f--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/609a0546634480569d2af85f>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://609a0d3b5493b66f8ae0e921--gamut-preview.netlify.app>https://609a0d3b5493b66f8ae0e921--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/609a0d3b5493b66f8ae0e921>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@4.12.1-alpha.e660de.0<br/>@codecademy/gamut-illustrations@0.14.2-alpha.e660de.0<br/>@codecademy/gamut-kit@0.3.89-alpha.e660de.0<br/>@codecademy/gamut-labs@15.8.2-alpha.e660de.0<br/>@codecademy/gamut-patterns@0.1.2-alpha.e660de.0<br/>@codecademy/gamut-styles@9.8.2-alpha.e660de.0<br/>@codecademy/gamut-tests@2.3.29-alpha.e660de.0<br/>@codecademy/gamut@32.15.3-alpha.e660de.0<br/>@codecademy/konamimojisplosion@0.2.29-alpha.e660de.0<br/>@codecademy/markdown-overrides@0.5.41-alpha.e660de.0<br/>@codecademy/styleguide@37.2.3-alpha.e660de.0<br/>@codecademy/variance@0.11.3-alpha.e660de.0<br/></pre>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@4.12.2-alpha.fb0deb.0<br/>@codecademy/gamut-illustrations@0.14.3-alpha.fb0deb.0<br/>@codecademy/gamut-kit@0.3.92-alpha.fb0deb.0<br/>@codecademy/gamut-labs@15.8.5-alpha.fb0deb.0<br/>@codecademy/gamut-patterns@0.1.3-alpha.fb0deb.0<br/>@codecademy/gamut-styles@9.9.1-alpha.fb0deb.0<br/>@codecademy/gamut-tests@2.3.30-alpha.fb0deb.0<br/>@codecademy/gamut@32.17.1-alpha.fb0deb.0<br/>@codecademy/konamimojisplosion@0.2.30-alpha.fb0deb.0<br/>@codecademy/markdown-overrides@0.5.42-alpha.fb0deb.0<br/>@codecademy/styleguide@37.4.1-alpha.fb0deb.0<br/>@codecademy/variance@0.11.3-alpha.fb0deb.0<br/></pre>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@4.12.2-alpha.205a30.0<br/>@codecademy/gamut-illustrations@0.14.3-alpha.205a30.0<br/>@codecademy/gamut-kit@0.3.95-alpha.205a30.0<br/>@codecademy/gamut-labs@15.8.8-alpha.205a30.0<br/>@codecademy/gamut-patterns@0.1.3-alpha.205a30.0<br/>@codecademy/gamut-styles@9.9.1-alpha.205a30.0<br/>@codecademy/gamut-tests@2.3.30-alpha.205a30.0<br/>@codecademy/gamut@32.18.3-alpha.205a30.0<br/>@codecademy/konamimojisplosion@0.2.30-alpha.205a30.0<br/>@codecademy/markdown-overrides@0.5.42-alpha.205a30.0<br/>@codecademy/styleguide@37.5.3-alpha.205a30.0<br/>@codecademy/variance@0.11.3-alpha.205a30.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://609afab6fcf7812a0dc38af5--gamut-preview.netlify.app>https://609afab6fcf7812a0dc38af5--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/609afab6fcf7812a0dc38af5>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://609b042abcd98238511cad44--gamut-preview.netlify.app>https://609b042abcd98238511cad44--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/609b042abcd98238511cad44>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@4.12.2-alpha.60fa78.0<br/>@codecademy/gamut-illustrations@0.14.3-alpha.60fa78.0<br/>@codecademy/gamut-kit@0.3.95-alpha.60fa78.0<br/>@codecademy/gamut-labs@15.8.8-alpha.60fa78.0<br/>@codecademy/gamut-patterns@0.1.3-alpha.60fa78.0<br/>@codecademy/gamut-styles@9.9.1-alpha.60fa78.0<br/>@codecademy/gamut-tests@2.3.30-alpha.60fa78.0<br/>@codecademy/gamut@32.18.3-alpha.60fa78.0<br/>@codecademy/konamimojisplosion@0.2.30-alpha.60fa78.0<br/>@codecademy/markdown-overrides@0.5.42-alpha.60fa78.0<br/>@codecademy/styleguide@37.5.3-alpha.60fa78.0<br/>@codecademy/variance@0.11.3-alpha.60fa78.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://609b5be54fadb9667ee89c60--gamut-preview.netlify.app>https://609b5be54fadb9667ee89c60--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/609b5be54fadb9667ee89c60>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@4.12.2-alpha.53a6e2.0<br/>@codecademy/gamut-illustrations@0.14.3-alpha.53a6e2.0<br/>@codecademy/gamut-kit@0.3.95-alpha.53a6e2.0<br/>@codecademy/gamut-labs@15.8.8-alpha.53a6e2.0<br/>@codecademy/gamut-patterns@0.1.3-alpha.53a6e2.0<br/>@codecademy/gamut-styles@9.9.1-alpha.53a6e2.0<br/>@codecademy/gamut-tests@2.3.30-alpha.53a6e2.0<br/>@codecademy/gamut@32.18.3-alpha.53a6e2.0<br/>@codecademy/konamimojisplosion@0.2.30-alpha.53a6e2.0<br/>@codecademy/markdown-overrides@0.5.42-alpha.53a6e2.0<br/>@codecademy/styleguide@37.5.3-alpha.53a6e2.0<br/>@codecademy/variance@0.11.3-alpha.53a6e2.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://609c2c3397638f0eecf64a1b--gamut-preview.netlify.app>https://609c2c3397638f0eecf64a1b--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/609c2c3397638f0eecf64a1b>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@4.12.2-alpha.fedd7c.0<br/>@codecademy/gamut-illustrations@0.14.3-alpha.fedd7c.0<br/>@codecademy/gamut-kit@0.3.95-alpha.fedd7c.0<br/>@codecademy/gamut-labs@15.8.8-alpha.fedd7c.0<br/>@codecademy/gamut-patterns@0.1.3-alpha.fedd7c.0<br/>@codecademy/gamut-styles@9.9.1-alpha.fedd7c.0<br/>@codecademy/gamut-tests@2.3.30-alpha.fedd7c.0<br/>@codecademy/gamut@32.18.3-alpha.fedd7c.0<br/>@codecademy/konamimojisplosion@0.2.30-alpha.fedd7c.0<br/>@codecademy/markdown-overrides@0.5.42-alpha.fedd7c.0<br/>@codecademy/styleguide@37.5.3-alpha.fedd7c.0<br/>@codecademy/variance@0.11.3-alpha.fedd7c.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://609c4b8d315c1e31493a06a9--gamut-preview.netlify.app>https://609c4b8d315c1e31493a06a9--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/609c4b8d315c1e31493a06a9>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@4.12.2-alpha.15d553.0<br/>@codecademy/gamut-illustrations@0.14.3-alpha.15d553.0<br/>@codecademy/gamut-kit@0.3.95-alpha.15d553.0<br/>@codecademy/gamut-labs@15.8.8-alpha.15d553.0<br/>@codecademy/gamut-patterns@0.1.3-alpha.15d553.0<br/>@codecademy/gamut-styles@9.9.1-alpha.15d553.0<br/>@codecademy/gamut-tests@2.3.30-alpha.15d553.0<br/>@codecademy/gamut@32.18.3-alpha.15d553.0<br/>@codecademy/konamimojisplosion@0.2.30-alpha.15d553.0<br/>@codecademy/markdown-overrides@0.5.42-alpha.15d553.0<br/>@codecademy/styleguide@37.5.3-alpha.15d553.0<br/>@codecademy/variance@0.11.3-alpha.15d553.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://609ca021f0663e4b734f8ad8--gamut-preview.netlify.app>https://609ca021f0663e4b734f8ad8--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/609ca021f0663e4b734f8ad8>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@4.12.2-alpha.3e80b8.0<br/>@codecademy/gamut-illustrations@0.14.3-alpha.3e80b8.0<br/>@codecademy/gamut-kit@0.3.95-alpha.3e80b8.0<br/>@codecademy/gamut-labs@15.8.8-alpha.3e80b8.0<br/>@codecademy/gamut-patterns@0.1.3-alpha.3e80b8.0<br/>@codecademy/gamut-styles@9.9.1-alpha.3e80b8.0<br/>@codecademy/gamut-tests@2.3.30-alpha.3e80b8.0<br/>@codecademy/gamut@32.18.3-alpha.3e80b8.0<br/>@codecademy/konamimojisplosion@0.2.30-alpha.3e80b8.0<br/>@codecademy/markdown-overrides@0.5.42-alpha.3e80b8.0<br/>@codecademy/styleguide@37.5.3-alpha.3e80b8.0<br/>@codecademy/variance@0.11.3-alpha.3e80b8.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://609cb164e5d1cd58c323192e--gamut-preview.netlify.app>https://609cb164e5d1cd58c323192e--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/609cb164e5d1cd58c323192e>Deploy Logs</a></p>

File: packages/gamut-labs/src/AppHeaderMobile/AppHeaderLinkMobile/index.tsx [hunk 1]
```diff
@@ -21,7 +21,7 @@ type AppHeaderLinkButtonProps = { topSeparator: boolean };
 
 const SeparatorOuter = styled(Box)<AppHeaderLinkButtonProps>`
   border-top: ${({ theme, topSeparator }) =>
-    topSeparator ? `1px solid ${theme.colors['gray-400']}` : ''};
+    topSeparator ? `1px solid ${theme.colors['gray-600']}` : ''};
```
Comment (JoshuaKGoldberg) [inline packages/gamut-labs/src/AppHeaderMobile/AppHeaderLinkMobile/index.tsx:24]: Could you split these cleanups into a separate PR please? This one's pretty huge as is.

**Comment by codecaaron** [comment]
Closing in favor of smaller PRs
